src/lib/stream/ftp.ts

```diff
@@ -0,0 +1,181 @@
+import FTP from 'ftp'; // Import the 'ftp' module for FTP operations
+import nodeCleanup from 'node-cleanup'; // Import the 'finalize' function from the 'node-cleanup' module
+import { Readable } from 'stream';
+
+interface FTPSettings {
+  host: string, // The FTP server host
+  port: number, // The FTP server port
+  username: string, // The FTP server username
+  password: string, // The FTP server password
+}
+
+interface FileInfo {
+  path: string,
+  name: string, // The name of the file
+  type: string, // The type of the file (e.g., file or directory)
+  size: number, // The size of the file in bytes
+  date: Date, // The last modified date of the file
+  target?: string, // The target of a symbolic link (optional)
+  stream?: Readable, // Stream to file (optional)
+}
+
+function joinPaths(basePath: string, fileName: string): string {
+  if (!basePath || !fileName) return basePath || fileName;
+  return `${basePath.replace(/\/+$/, '')}/${fileName.replace(/^\/+/, '')}`;
+}
+
+class FtpClient {
+  private client; // Private property to store the FTP client instance
+
+  // One way to handle this is to use a static property to track if cleanup has been registered
+  static cleanupRegistered = false;
+
+  // Use a static method to handle cleanup
+  static handleCleanup(client: FTP) {
+    if (client && client.connected) {
+      client.end();
+    }
+  }
+
+  constructor(
+    private ftpSettings: FTPSettings, // Constructor parameter to store the FTP settings
+  ) {
+    this.client = new FTP(); // Create a new FTP client instance
+    if (!FtpClient.cleanupRegistered) {
+      nodeCleanup(() => {
```

Comment:
You can use `process.on('exit', callback)` instead of bringing in the `node-cleanup` module.

Reply:
it does seem like the only reason to use a package here would be to catch certain codes